### ISSUE TITLE
All ScrollDown and ScrollDownTo functions are broken for iOS in Xamarin.UITest version 3.2.4 - 3.2.6

### ISSUE THREAD

**Issue by Moshex** [microsoft/appcenter#2349]
When updating to Xamarin.UITest version 3.2.4 - 3.2.6 the ScrollDown and ScrollDownTo functions are not working for iOS. When using 
app.ScrollDown("{element}") or app.ScrollDownTo("{element}") Xamarin.UITest will display the foilowing error:

```
System.Exception : Error while performing ScrollDown(Marked("{element}"), Gesture, 0.67, 500, True)
  ----> System.Exception : GestureBasedScroll: Unable to swipe in direction Down
```

This issue only happens on iOS and ScrollDown is working just fine for Android. Also, the version of iOS device that this was tested on to see the issue, was iOS version  14.8. 

When downgrading the Xamarin.UITest version to 3.2.3 the issue no longer occurs, that I have already tested on my end.
Thank you in advance for looking into this issue.

**Comment by konstantin-bobkov** [comment]
@Moshex, hello! Have you experienced this issue with Xamarin.UITest 3.2.5?

**Comment by Moshex** [comment]
Not sure, just saw the update. I will test and find out

**Comment by Dmbandit** [comment]
Looks like this might still be an issue. I got the same system exception with UITest 3.2.5. 

> iOS test running **Xamarin.UITest version: 3.2.5**
Using TEST_CLOUD configuration override
Running in Test Cloud, on device with iOS8 or higher.
Test server version: 0.21.8........
......Scrolling down
Error while performing ScrollDown(null, Gesture, 1, 500, True)
**Exception: System.Exception: GestureBasedScroll: Unable to swipe in direction Down**  
at Xamarin.UITest.iOS.iOSGestures.GestureBasedScroll (Xamarin.UITest.Queries.AppQuery withinQuery, Xamarin.UITest.ScrollDirection scrollDirection, System.Double swipePercentage, System.Int32 swipeSpeed, System.Boolean withInertia) [0x0012e] in <5b03a2e60ad4475bb53699724c965d8e>:0  
at Xamarin.UITest.iOS.iOSGestures.Scroll (Xamarin.UITest.Queries.AppQuery withQuery, Xamarin.UITest.ScrollDirection direction, Xamarin.UITest.ScrollStrategy strategy, System.Double swipePercentage, System.Int32 swipeSpeed, System.Boolean withInertia) [0x0000a] in <5b03a2e60ad4475bb53699724c965d8e>:0  
at Xamarin.UITest.SharedApp.ScrollDown (System.Double swipePercentage, System.Int32 swipeSpeed, System.Func`2[T,TResult] withinQuery, Xamarin.UITest.ScrollStrategy strategy, System.Boolean withInertia) [0x00019] in <5b03a2e60ad4475bb53699724c965d8e>:0  
at Xamarin.UITest.iOS.iOSApp+<>c__DisplayClass93_0.<ScrollDown>b__0 () [0x00000] in <5b03a2e60ad4475bb53699724c965d8e>:0  
at Xamarin.UITest.Utils.ErrorReporting.With (System.Action func, System.Object[] args, System.String memberName) [0x00006] in <5b03a2e60ad4475bb53699724c965d8e>:0

Can anyone else verify?

**Comment by Moshex** [comment]
@Dmbandit @konstantin-bobkov 
after my testing, yes, version 3.2.5 still has the same bug and it's not a usable version.

**Comment by wakeboardfit** [comment]
I found this as well.  Updated to 3.2.6 today and experimented with Android vs iOS running in App Center.  It appears the behavior of the scrolling is broken across both platforms depending on which you use, if you use string vs query, and I have not found a cross-section of where both work using the same call.
A couple examples:
app.ScrollTo("<element>");
Android - Works
iOS - Doesn't Work

app.ScrollDownTo(c => c.Marked("<element>"));
Android - Doesn't Work
iOS - Works

app.ScrollDown(c => c.Marked("<element>"));
Android - Doesn't Work
iOS - Works

I'm going to play with maybe using a common function that switches depending on platform but it seems the behavior should be consistent no matter which you use.  Or, at least have one that works on both platforms.

**Comment by Dmbandit** [comment]
> I found this as well. Updated to 3.2.6 today and experimented with Android vs iOS running in App Center. It appears the behavior of the scrolling is broken across both platforms depending on which you use, if you use string vs query, and I have not found a cross-section of where both work using the same call. A couple examples: app.ScrollTo(""); Android - Works iOS - Doesn't Work
> 
> app.ScrollDownTo(c => c.Marked("")); Android - Doesn't Work iOS - Works
> 
> app.ScrollDown(c => c.Marked("")); Android - Doesn't Work iOS - Works
> 
> I'm going to play with maybe using a common function that switches depending on platform but it seems the behavior should be consistent no matter which you use. Or, at least have one that works on both platforms.

Thanks for that info

**Comment by wakeboardfit** [comment]
Threw a quick thing together this morning so I could move on if you want to use something similar.  Just going to use this temporarily until the bugs in UITest are fixed.  This is working in my tests now on Android and iOS.

public static void DoScroll(IApp app, string element)
        {
            if (Global.Platform == Platform.Android)
            {
                app.ScrollTo(element);
            }
            else if (Global.Platform == Platform.iOS)
            {
                app.ScrollDownTo(c => c.Marked(element));
            }
            else
            {
                throw new Exception("No Platform Defined");
            }
        }

**Comment by Dmbandit** [comment]
> Threw a quick thing together this morning so I could move on if you want to use something similar. Just going to use this temporarily until the bugs in UITest are fixed. This is working in my tests now on Android and iOS.
> 
> public static void DoScroll(IApp app, string element) { if (Global.Platform == Platform.Android) { app.ScrollTo(element); } else if (Global.Platform == Platform.iOS) { app.ScrollDownTo(c => c.Marked(element)); } else { throw new Exception("No Platform Defined"); } }

I was going to do something similar but now you have done the thinking on my behalf. I appreciate you!

**Comment by Moshex** [comment]
@wakeboardfit I am not sure that is a working solution since:
app.ScrollDownTo(c => c.Marked("")); 
app.ScrollDown(c => c.Marked("")); 
work diffrently when scolling, both methods should work for both platforms as intended. 
right now the latest usable version of the Xamrian UI test is version 3.2.3 where all the scroll-down methods work as intended. 
I hope to see a fix for this issue very soon.

**Comment by AlexeyBudynkov** [comment]
@Moshex Hello. I saw this line in your logs 
_Test server version: 0.21.8_
For now the current stable version is 0.23.1.
Could you try to update it and then to use scroll methods again?

**Comment by AlexeyBudynkov** [comment]
> @AlexeyBudynkov in this thread what logs do you see that I posted my "Test server version: 0.21.8"? The exceptions are across the board with no relevance to the test server. also, the issue happens locally not just in app-center @AlexeyBudynkov are you saying that on version 3.2.6 all scroll down functions are working for you as intended for all ios versions?

It's not just Xamarin.UITest is responsible for some gestures.
Did you include calabash.framework in your project when you started to use Xamarin.UITest?

**Comment by Dmbandit** [comment]
> @Moshex Hello. I saw this line in your logs _Test server version: 0.21.8_ For now the current stable version is 0.23.1. Could you try to update it and then to use scroll methods again?
> 
> If you use Xamarin.Forms for your application, then you should update Test Cloud Agent nuget package. If you use a native application, then you should update calabash framework.

I can confirm this makes no difference. Still broken after updating.

**Comment by Dmbandit** [comment]
I was still having some issues with iOS but I managed to get things working (on UI Test version 3.2.5) after some REPLing.
If you give the actual ListView/ScrollView that you are trying to scroll inside an AutomationId then the following works on iOS:

app.ScrollDownTo("MarkedElement", "WithinElement");

Where MarkedElement is the item you are scrolling to and WithinElement is the AutomationId of the ScrollView or ListView. I didn't know about this overload in the past and it seems to work well on iOS, even just using the item Label Text + Automation Id as strings. Maybe it will help someone.

ScrollDownTo("ScrollTargetLabel") does not work.
ScrollDownTo("ScrollTargetLabel", "ListViewAutomationId") works for me.

On Android the regular ScrollTo("ScrollTargetLabel") method seems to works fine.

**Comment by AlexeyBudynkov** [comment]
> > @Moshex Hello. I saw this line in your logs _Test server version: 0.21.8_ For now the current stable version is 0.23.1. Could you try to update it and then to use scroll methods again?
> > If you use Xamarin.Forms for your application, then you should update Test Cloud Agent nuget package. If you use a native application, then you should update calabash framework.
> 
> I can confirm this makes no difference. Still broken after updating.

You mean, that you have updated calabash.framework in your iOS project AND Xamarin.UITest version for your test project?

**Comment by Dmbandit** [comment]
> > > @Moshex Hello. I saw this line in your logs _Test server version: 0.21.8_ For now the current stable version is 0.23.1. Could you try to update it and then to use scroll methods again?
> > > If you use Xamarin.Forms for your application, then you should update Test Cloud Agent nuget package. If you use a native application, then you should update calabash framework.
> > 
> > 
> > I can confirm this makes no difference. Still broken after updating.
> 
> You mean, that you have updated calabash.framework in your iOS project AND Xamarin.UITest version for your test project?

Correct. ScrollDownTo is still broken as reported by this issue.

**Comment by Dmbandit** [comment]
> I was still having some issues with iOS but I managed to get things working (on UI Test version 3.2.5) after some REPLing. If you give the actual ListView/ScrollView that you are trying to scroll inside an AutomationId then the following works on iOS:
> 
> app.ScrollDownTo("MarkedElement", "WithinElement");
> 
> Where MarkedElement is the item you are scrolling to and WithinElement is the AutomationId of the ScrollView or ListView. I didn't know about this overload in the past and it seems to work well on iOS, even just using the item Label Text + Automation Id as strings. Maybe it will help someone.
> 
> ScrollDownTo("ScrollTargetLabel") does not work. ScrollDownTo("ScrollTargetLabel", "ListViewAutomationId") works for me.
> 
> On Android the regular ScrollTo("ScrollTargetLabel") method seems to works fine.

Actually it looks like upgrading to the latest Test Cloud Agent has broken the above also. I downgraded back to the version I was on 0.22.1 and ScrollDownTo("ScrollTargetLabel", "ListViewAutomationId") started working again.

**Comment by AlexeyBudynkov** [comment]
> > I was still having some issues with iOS but I managed to get things working (on UI Test version 3.2.5) after some REPLing. If you give the actual ListView/ScrollView that you are trying to scroll inside an AutomationId then the following works on iOS:
> > app.ScrollDownTo("MarkedElement", "WithinElement");
> > Where MarkedElement is the item you are scrolling to and WithinElement is the AutomationId of the ScrollView or ListView. I didn't know about this overload in the past and it seems to work well on iOS, even just using the item Label Text + Automation Id as strings. Maybe it will help someone.
> > ScrollDownTo("ScrollTargetLabel") does not work. ScrollDownTo("ScrollTargetLabel", "ListViewAutomationId") works for me.
> > On Android the regular ScrollTo("ScrollTargetLabel") method seems to works fine.
> 
> Actually it looks like upgrading to the latest Test Cloud Agent has broken the above also. I downgraded back to the version I was on 0.22.1 and ScrollDownTo("ScrollTargetLabel", "ListViewAutomationId") started working again.

Thanks for your investigation, it will help to fix it

**Comment by joshardt** [comment]
Still reproducible with the newest version.

**Comment by Moshex** [comment]
@joshardt does version 3.2.7 still have the same issue?

**Comment by Banley** [comment]
> @joshardt does version 3.2.7 still have the same issue?

It seemed to me like scrolling was fixed in 3.2.7.

**Comment by joshardt** [comment]
I've tested with version 3.2.6.
For a real test I need to run the UI test on AppCenter. I for my colleague @ARomoth will do the test and let you know.

**Comment by Zure1** [comment]
> I've tested with version 3.2.6. For a real test I need to run the UI test on AppCenter. I or my colleague @ARomoth will do the test and let you know.

Tested with version 3.2.7.
The ScrollDownTo method works as expected locally and in AppCenter.

**Comment by Moshex** [comment]
I have tested version 3.2.7, and it is working as expected on that version for me as well.

**Comment by wakeboardfit** [comment]
I will test on 3.2.7 as well as I was still having to use the helper method posted above depending on OS.

**Comment by IlyaBausovAkvelon** [comment]
Hi, All. ScrollDown issue was fixed in 3.2.7 version. Please consider upgrading to this version.

**Comment by generator87** [comment]
Tested with version 3.2.8 and app.ScrollDown()  does not work. Nothing changes on the mobile app when method is executed.

**Comment by generator87** [comment]
> > Tested with version 3.2.8 and app.ScrollDown() works on Android when strategy is Gesture, not Auto:
> > app.ScrollDown(strategy:ScrollStrategy.Gesture)
> > have not tried the remaining strategies.
> 
> @generator87, could you please confirm that initial issue is gone?

Unfortunately I don't have any iOS device at the moment. I hope that someone else will be able to confirm that initial issue is gone.

**Comment by wakeboardfit** [comment]
I checked 3.2.8 and was able to remove my custom handling of scrolling between iOS and Android.  ScrollDownTo seems to be working better now.

**Comment by wakeboardfit** [comment]
Thanks for the suggestion.  I found, if I swap ScrollDownTo with ScrollTo, that it works on Android correctly.  So, it seems it may be a difference in how those two behave where it's either how far it scrolls or the speed or how it recognizes when something appears.  I have an issue on iOS but it seems to be a visible vs clickable on screen size.  
![image](https://user-images.githubusercontent.com/8449950/179815276-ffde2569-21bb-4f54-a897-f4f09acd855e.png)
In this case, the field is on the screen so it doesn't try to scroll but it's not on the screen enough to click it evidently: ----> System.TimeoutException : Timed out waiting for keyboard.

**Comment by Banley** [comment]
I have seen 2 similar issues as well in iOS. Some ideas below. With the rate the issues are being addressed, some workarounds may be in order.

With the "object present but half off the screen" issue, you may try to tap a corner or side of the control (default is to tap dead center, I believe).
"Timed out waiting for keyboard" -- MIGHT not be a scroll issue. I see this regularly and I'm still not sure how to workaround: Sometimes a textbox control will suddenly be unresponsive, the keyboard will not appear when tapping the textbox. This seems to be random as a subsequent run will most likely pass. Others have seen this too, but not sure it's logged solitarily [the following item addressed it briefly but is closed: https://github.com/microsoft/appcenter/issues/2295].

**Comment by Moshex** [comment]
@wakeboardfit , I had an issue like the one you are experiencing however I have been seeing this issue for 2 to 3 years, and not much of a resolution. This may be an issue that does not have to do with version 3.2.8.

here is the issue I reported:

https://github.com/microsoft/appcenter/issues/2201

**Comment by wakeboardfit** [comment]
Yeah, def seems like there are some shortcomings in behavior.  Specifically around iOS handling of things.

**Comment by wakeboardfit** [comment]
Figured I'd share if anyone else needs it.  To get scrolling to work the same on Android and iOS, I'm doing the following.  This handles the different behavior in the ScrollXYZ functions as well as when the field is on screen but not actually clickable by moving the field to be fully displayed.
`public static void DoScroll(IApp app, string element, string scrollPane="")
        {            
            if (Global.Platform == Platform.Android)
            {
                app.ScrollTo(element, timeout: new TimeSpan(0, 1, 0));                
            }
            else if (Global.Platform == Platform.iOS)
            {
                app.ScrollDownTo(c => c.Marked(element), timeout: new TimeSpan(0, 1, 0));
            }

            var content = app.Query(element).First();
            var bottomLeftYCoordinateContent = content.Rect.Y + content.Rect.Height;

            var window = app.Query(scrollPane).First();
            var bottomLeftYCoordinate = window.Rect.Y + window.Rect.Height;

            if (bottomLeftYCoordinateContent > bottomLeftYCoordinate)
            {
                app.ScrollDown();
                DoScroll(app, element, scrollPane);
            }
        }`

**Comment by wakeboardfit** [comment]
I don't think this issue has been resolved nor has the issue of correct scroll behavior in iOS as linked in this thread.  I still am running custom handlers for improved behavior to solve these issues.

**Comment by wakeboardfit** [comment]
I'm not sure if this is what you want:
appcenter version 2.10.10

I see no other agent version on the AppCenter site.

**Comment by MrAusnadian** [comment]
Bot bad. Issue not fixed

**Comment by IlyaBausovAkvelon** [comment]
@MrAusnadian TestCloudAgent is the NuGet package you should be using to run the Xamarin.UITest. It should be included in your main app project. What version do you use?

**Comment by wakeboardfit** [comment]
Running UITest 3.2.9 and have now seen that I used to use ScrollTo in Android and that worked, now I need to use ScrollDownTo.

**Comment by rafageist** [comment]
ScrollDown, ScrollTo, ScrollDownTo, .... and DragAndDrop, DragCoordinates .... Nothing works on iOS with Xamarin UITest 4.3.3, only Taps. Is there any solution for this?

I've tried everything and still nothing. Even simulate scrolling with Drag. All my solutions work on Android, but not on iOS in AppCenter.

**Comment by IlyaBausovAkvelon** [comment]
Hi @rafageist,
Thank you for reporting the issue.
What iOS version do you use for testing?

**Comment by Moshex** [comment]
@rafageist 
I am not sure about version 4.3.3 since I have written up a whole report on how that version is very unstable and does not work as expected. Do you have scroll issues on version 4.3.2? if yes I am not seeing the same issue on my end. However, since it does not work on a physical device I have only tested it on an iOS simulator.

**Comment by rafageist** [comment]
I've conducted extensive testing on AppCenter, specifically focusing on automation features, using the following iOS devices:

Apple iPhone 8 Plus

iOS Version: 16.7.4
Release Date: September 2017

Apple iPhone X

iOS Version: 16.7.4
Release Date: October 2017

Regrettably, none of the automation features—ScrollDown, ScrollTo, ScrollDownTo, DragAndDrop, DragCoordinates—seem to be functioning as expected on these iOS devices with Xamarin UITest version 4.3.3. It's noteworthy that all these features work seamlessly on Android devices.

Additionally, I have tested with various versions of Xamarin UITest, including both the 4.x.x and 3.x.x series, but unfortunately, the issue persists on iOS.

In my exploration for solutions, I also came across a note in the [official documentation](https://learn.microsoft.com/en-us/appcenter/test-cloud/frameworks/uitest/features/gestures) that mentions a bug in the iOS simulator: **"Because of a bug in the iOS simulator, flick/swipe doesn't work in scroll views such as UITableView or UIScrollView."** (I still wonder if AppCenter uses physical devices)

If there are any known issues or specific configurations for iOS that I might be overlooking, your guidance would be immensely appreciated. Thank you for your continued assistance.
